File: src/Algorithm/LeetCode/LeetCode_541_Reverse_String_II/Solution_MyOwn.java
package Algorithm.LeetCode.LeetCode_541_Reverse_String_II;

class Solution_MyOwn {
    public String reverseStr(String s, int k) {
        int i = 0;

        char[] chars = s.toCharArray();

        while (i < chars.length) {
            reverse(chars, i, Math.min(i + k, chars.length) - 1);
            i += 2 * k;
        }

        return new String(chars);

    }

    private void reverse(char[] chars, int start, int end) {
        int i = start, j = end;

        while (i <= j) {
            char temp = chars[i];
            chars[i] = chars[j];
            chars[j] = temp;
            i++;
            j--;
        }
    }
}

File: src/Algorithm/LeetCode/LeetCode_541_Reverse_String_II/Solution_MyOwn.py
from typing import List


def reverse(s: List[str], left: int, right: int):
    while left < right:
        tmp = s[left]
        s[left] = s[right]
        s[right] = tmp

        left += 1
        right -= 1

    return s


class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        i = 0
        res = list(s)

        while i < len(s):
            left = i
            right = min(i + k - 1, len(s) - 1)

            reverse(res, left, right)

            i += 2 * k

        return ''.join(res)
